fuel-vm/src/predicate.rs

```diff
@@ -59,108 +61,205 @@ fn from_tx_works() {
+            rng.gen(),
+            predicate.clone(),
+            predicate_data.clone(),
+        );
+
+        let c = Input::message_data_predicate(
+            rng.gen(),
+            rng.gen(),
+            rng.gen(),
+            rng.gen(),
+            rng.gen(),
+            vec![0xff; 10],
+            predicate.clone(),
+            predicate_data,
+        );
+
+        let inputs = vec![a, b, c];
+
+        for i in inputs {
+            let tx = TransactionBuilder::script(vec![], vec![])
+                .add_input(i)
+                .add_random_fee_input()
+                .finalize_checked_basic(height);
+
+            // assert invalid idx wont panic
+            let idx = 1;
+            let tx_offset = TxParameters::DEFAULT.tx_offset();
+            let runtime = RuntimePredicate::from_tx(tx.as_ref(), tx_offset, idx);
+
+            assert!(runtime.is_none());
+
+            // fetch the input predicate
+            let idx = 0;
+            let runtime = RuntimePredicate::from_tx(tx.as_ref(), tx_offset, idx)
+                .expect("failed to generate predicate from valid tx");
+
+            assert_eq!(idx, runtime.idx());
+
+            let mut interpreter =
+                Interpreter::with_storage(PredicateStorage, InterpreterParams::default());
+
+            assert!(interpreter
+                .init_predicate(
+                    Context::PredicateVerification {
+                        program: Default::default()
+                    },
+                    tx.transaction().clone(),
+                    tx.transaction().limit()
+                )
+                .is_ok());
+
+            let pad = bytes::padded_len(&predicate) - predicate.len();
+
+            // assert we are testing an edge case
+            assert_ne!(0, pad);
+
+            let padded_predicate: Vec<u8> = predicate
+                .iter()
+                .copied()
+                .chain(iter::repeat(0u8).take(pad))
+                .collect();
+
+            let program = runtime.program();
+            let program = &interpreter.memory()[program.usizes()];
+
+            // assert the program in the vm memory is the same of the input
+            assert_eq!(program, &padded_predicate);
+        }
+    }
+
+    /// Verifies the runtime predicate validation rules outlined in the spec are actually
+    /// validated https://github.com/FuelLabs/fuel-specs/blob/master/src/fuel-vm/index.md#predicate-verification
+    #[test]
+    fn inputs_are_validated() {
+        let rng = &mut StdRng::seed_from_u64(2322u64);
+
+        let height = 1.into();
+        let predicate_data =
+            b"If you think it's simple, then you have misunderstood the problem."
+                .to_vec();
+
+        macro_rules! predicate_input {
+            ($predicate:expr) => {{
+                let predicate: Vec<u8> = $predicate.into_iter().collect();
+                let owner = Input::predicate_owner(&predicate, &ChainId::default());
```

Comment:
I thought that this issue was related to `message_data_predicate` and `message_coin_predicate` inputs=D

Reply:
Oops fixed the wrong thing =D. Moving back to a draft so I'll add tests fro those as well.